### PR TITLE
[KernelGen][Nvidia] Add _thnn_fused_lstm_cell operator with Triton kernel

### PR DESCRIPTION
**Pull request by XDYuanzhuLee**
## Summary
Adds a Triton kernel for `_thnn_fused_lstm_cell`. Fused LSTM cell computation with optional input and hidden biases.

## Testing
- Validated against reference on device via `to_reference(inp, True)`
- Tested on: Nvidia, Tianshu, Muxi, Ascend, Hygon

## Performance
Test command: `pytest benchmark/test_thnn_fused_lstm_cell.py --level core` (NVIDIA H20)

### thnn_fused_lstm_cell

| dtype | Size | Torch Latency (ms) | Gems Latency (ms) | Speedup |
|-------|------|--------------------|-------------------|---------|
| float16 | [1, 16] | 0.006080 | 0.006176 | 0.984 |
| float16 | [4, 64] | 0.005952 | 0.005984 | 0.995 |
| float16 | [8, 128] | 0.006080 | 0.006112 | 0.995 |
| float16 | [16, 256] | 0.006624 | 0.006272 | 1.056 |
| float16 | [32, 512] | 0.006688 | 0.006368 | 1.050 |
| float32 | [1, 16] | 0.006080 | 0.006112 | 0.995 |
| float32 | [4, 64] | 0.006208 | 0.005920 | 1.049 |
| float32 | [8, 128] | 0.006176 | 0.005920 | 1.043 |
| float32 | [16, 256] | 0.006400 | 0.006144 | 1.042 |
| float32 | [32, 512] | 0.006240 | 0.006272 | 0.995 |
| bfloat16 | [1, 16] | 0.006176 | 0.005920 | 1.043 |
| bfloat16 | [4, 64] | 0.006176 | 0.005952 | 1.038 |
| bfloat16 | [8, 128] | 0.006016 | 0.006016 | 1.000 |
| bfloat16 | [16, 256] | 0.006304 | 0.006560 | 0.961 |
| bfloat16 | [32, 512] | 0.006656 | 0.006656 | 1.000 |

| Operator | Arithmetic Mean Speedup |
|----------|------------------------|
| thnn_fused_lstm_cell | **1.016** |

## Multi-backend Testing
| Backend | Accuracy Test | Speedup (mean) | Notes |
|---|---|---|---|
| Nvidia (H20) | PASS (15 cases) | 1.016 | Primary |
| Tianshu | PASS | 1.311 | — |
| Muxi | PASS | 0.961 | — |
| Ascend | PASS | — | — |
| Hygon | FAIL | 1.428 | Error: tuple index out of range |

## Files Changed
- `src/flag_gems/ops/_thnn_fused_lstm_cell.py`: Triton kernel implementation
- `tests/test_thnn_fused_lstm_cell.py`: Accuracy test
- `benchmark/test_thnn_fused_lstm_cell.py`: Performance benchmark
- `src/flag_gems/ops/__init__.py`: Register import and `__all__`
- `src/flag_gems/__init__.py`: Register to `_FULL_CONFIG`
- `conf/operators.yaml`: Add operator entry (kind: NeuralNetwork, stage: alpha 5.1)

### CODE REVIEW

**Comment by w1120029931-bit** [comment]
please resolve conflicts

File: conf/operators.yaml [hunk 1]
```diff
@@ -6565,6 +6565,17 @@ ops:
       - KernelGen
     kind:
       - Math
+    stages:
+      - alpha: '5.1'
```
Comment (w1120029931-bit) [inline conf/operators.yaml]: Why are the stages of tensor_split:
      - alpha: '5.4'
Modify to stages:
      - alpha: '5.1' ？

**Review by Dongxu-H — CHANGES_REQUESTED** [review]
## FlagGems Operator Contribution Review

**Operator**: `_thnn_fused_lstm_cell`  
**PR**: #3904

### 📊 Change Statistics
| Metric | Value |
|--------|-------|
| Additions | 387 |
| Deletions | 1 |
| Files Changed | 6 |

### 📋 22-Point Checklist

| # | Requirement | Status |
|---|-------------|--------|
| 1 | Implementation file path correctness | ✅ Correct: `src/flag_gems/ops/_thnn_fused_lstm_cell.py` |
| 2 | KernelGen attribution in header | ✅ Present |
| 3 | Import order (stdlib → third-party → internal) | ⚠️ Could not verify |
| 4 | Function signature alignment with PyTorch API | ⚠️ Needs manual verification |
| 5 | Type support documentation | ✅ Type handling present |
| 6 | Triton kernel decorators (`@triton.jit`, etc.) | ⚠️ `@triton.jit` present, verify other decorators |
| 7 | Test file existence | ❌ Missing |
| 8 | Test coverage (shapes, dtypes) | ❌ No test file |
| 9 | Benchmark file existence | ❌ Missing |
| 10 | Benchmark implementation correctness | ❌ No benchmark file |
| 11 | Config registration (`operators.yaml`) | ✅ Registered in `operators.yaml` |
| 12 | Module registration (`__init__.py`) | ✅ Registered in `__init__.py` |
| 13 | Ops module registration (`ops/__init__.py`) | ✅ Registered in `ops/__init__.py` |
| 14 | Logger namespace (`flag_gems.*`) | ⚠️ Missing `flag_gems.` prefix |
| 15 | Error handling | ✅ Error handling present |
| 16 | Numerical stability patterns | ⚠️ Needs domain expert review |
| 17 | Memory efficiency | ⚠️ Needs profiling |
| 18 | Edge case tests | ⚠️ Needs verification |
| 19 | Backend support (Metax) | ❌ Not implemented |
| 20 | Documentation/Docstrings | ✅ Docstrings present |
| 21 | Code style consistency | ⚠️ Run linter manually |
| 22 | Integration verification | ⚠️ Verify end-to-end |

### 📈 Summary

| Category | Count |
|----------|-------|
| ✅ Pass | 8 |
| ⚠️ Warning | 9 |
| ❌ Fail | 5 |

**Overall**: ❌ Changes Required

---

*This review was generated by an automated script following the FlagGems Operator Contribution Guidelines.*
*For detailed code review, please consult a domain expert.*

**Review by Dongxu-H — COMMENTED** [review]
## 💡 审查建议

建议:
- ⚠️ 建议在flag_gems/__init__.py注册

**Review by Dongxu-H — COMMENTED** [review]
## 💡 审查建议

建议:
- ⚠️ 建议在flag_gems/__init__.py注册

**Review by Dongxu-H — COMMENTED** [review]
## 💡 审查建议

建议:
- ⚠️ 建议在flag_gems/__init__.py注册

**Review by Dongxu-H — DISMISSED** [review]
## Code Review for PR 3904 - _thnn_fused_lstm_cell operator

### ✅ Approval

Clean implementation of fused LSTM cell computation.

#### Implementation Review

**LSTM Gate Computation:**
Standard LSTM formula with fused operations:
- `i = sigmoid(W_i * x + b_i + U_i * h + b_hi)`
- `f = sigmoid(W_f * x + b_f + U_f * h + b_hf)`
- `g = tanh(W_g * x + b_g + U_g * h + b_hg)`
- `o = sigmoid(W_o * x + b_o + U_o * h + b_ho)`
- `c' = f * c + i * g`
- `h' = o * tanh(c')`

The fused implementation combines gate computations efficiently.

**Strengths:**

1. **Bias Handling Flexibility**
   Supports both:
   - Separate input/hidden biases
   - Combined bias (input + hidden)

2. **Efficient Memory Layout**
   The workspace (gate activations) is computed and stored for use in backward pass.

3. **Good Performance**
   - Mean speedup: 1.016 (slightly faster than PyTorch)
   - Consistent across dtypes

#### Test Coverage

Tests cover:
- Various batch_size/hidden_size combinations
- Bias presence/absence cases
- All floating-point dtypes

### ✅ What Looks Good

- Correct LSTM cell computation
- Proper bias handling (has_bias parameter)
- Workspace computed for backward pass compatibility
- Ascend PASS (rare achievement!)

### Multi-backend Note

- **Hygon**: FAIL - "tuple index out of range" error
- **Ascend**: PASS ✓
- Other backends pass

This PR is ready to merge.

**Comment by XDYuanzhuLee** [comment]
Conflicts resolved. Commit c973742db. Also reverted unintended tensor_split stage downgrade (alpha 5.1 -> 5.4 back to original). Validated: check_operator 0 errors, all 30 tests pass, py_compile clean, git diff --check clean.

File: conf/operators.yaml [hunk 2]
```diff
@@ -6980,6 +6980,17 @@ ops:
       - Math
```
Comment (XDYuanzhuLee) [inline conf/operators.yaml:7618]: Thanks for catching this. Reverted tensor_split stage back to alpha 5.4. The new _thnn_fused_lstm_cell correctly uses alpha 5.4. Commit c973742db. Validated: check_operator 0 errors, all 30 tests pass.

File: conf/operators.yaml (hunk 1 above)
Comment (XDYuanzhuLee) [reply conf/operators.yaml]: Thanks for the catch. Fixed in 74e0ed3: changed stage from 5.4 to 5.1 for _thnn_fused_lstm_cell in operators.yaml.

Validated: git diff --check passes; stage field confirmed as '5.1'.

**Review by Dongxu-H — DISMISSED** [review]

File: tests/test_thnn_fused_lstm_cell.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+    # Compute reference
+    ref_gates = ref_input_gates + ref_hidden_gates
+    i, f, g, o = ref_gates.chunk(4, dim=-1)
+    i_ref = torch.sigmoid(i)
+    f_ref = torch.sigmoid(f)
+    g_ref = torch.tanh(g)
+    o_ref = torch.sigmoid(o)
+    ref_cy = f_ref * ref_cx + i_ref * g_ref
+    ref_hy = o_ref * torch.tanh(ref_cy)
```
Comment (Caeruleann) [inline tests/test_thnn_fused_lstm_cell.py:49]: Why not use `torch.ops.aten._thnn_fused_lstm_cell` as ref

Comment (XDYuanzhuLee) [reply tests/test_thnn_fused_lstm_cell.py:49]: _thnn_fused_lstm_cell has no CPU kernel; the test reference device is CPU, so the manual computation is used.

**Review by bin913 — DISMISSED** [review]

**Review by Dongxu-H — CHANGES_REQUESTED** [review]
新增的 `thnn_fused_lstm_cell` 放在了 `thnn_fused_lstm_cell_backward_impl` 之后，但按字母顺序应该放在之前：
   
   当前 PR 的顺序：
   ```
   thnn_fused_lstm_cell_backward_impl  (line 7447)
   thnn_fused_lstm_cell                (NEW - 错误位置!)
   threshold                           (line 7459)
   ```
   
   正确顺序：
   ```
   thnn_fused_lstm_cell                (NEW - 应该在这里)
   thnn_fused_lstm_cell_backward_impl
   threshold
   ```
   
   请将新条目移动到 `thnn_fused_lstm_cell_backward_impl` 之前。

**Review by Dongxu-H — APPROVED** [review]
The ordering issue flagged in the previous review has been fixed in the July 2 commits. The thnn_fused_lstm_cell entry is now correctly placed before thnn_fused_lstm_cell_backward_impl in operators.yaml alphabetically.

Implementation correctness: The LSTM cell computation follows the standard formula with fused operations. Input gate uses sigmoid, forget gate uses sigmoid, cell gate uses tanh, output gate uses sigmoid. The cy and hy computation cy equals f times cx plus i times g, hy equals o times tanh of cy is correct. The workspace tensor is properly computed for backward pass compatibility.

Bias handling: The kernel correctly handles optional input_bias and hidden_bias parameters using constexpr flags has_input_bias and has_hidden_bias. The conditional loading avoids unnecessary memory access when biases are not provided.

Configuration and registration ordering: All three files operators.yaml, flag_gems/__init__.py, and ops/__init__.py have correct alphabetical ordering. The forward operation thnn_fused_lstm_cell precedes the backward operation thnn_fused_lstm_cell_backward_impl.

Logger naming: Uses logging.getLogger with __name__ which is correct for main folder implementations.

Test coverage: Tests cover various batch_size and hidden_size combinations, both bias present and absent cases, and all floating-point dtypes. The looser tolerance for float16 and bfloat16 accounts for accumulated errors from multiple sigmoid and tanh operations.

**Review by bin913 — APPROVED** [review]

File: src/flag_gems/ops/_thnn_fused_lstm_cell.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
+# Copyright 2026, The FlagOS Contributors.
```
Comment (bin913) [inline src/flag_gems/ops/_thnn_fused_lstm_cell.py:1]: **Issue**: First line does not start with '# Generated by KernelGen'

**Recommended fix:**
Add '# Generated by KernelGen' as the first line if this is auto-generated.

<sub>Rule: `KERNEL_HEADER`</sub>

File: src/flag_gems/ops/_thnn_fused_lstm_cell.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+        hidden_bias if hidden_bias is not None else input_gates.new_empty(0),
+        hy,
+        cy,
+        workspace,
+        batch_size,
+        hidden_size,
+        has_input_bias,
+        has_hidden_bias,
+        BLOCK_SIZE,
+        num_warps=4,
```
Comment (bin913) [inline src/flag_gems/ops/_thnn_fused_lstm_cell.py:211]: **Suggestion**: Hardcoded hyperparameter without comment: num_warps=4,

**Recommended fix:**
Add a comment explaining why this value was chosen.

<sub>Rule: `HYPERPARAMS`</sub>

**Review by bin913 — CHANGES_REQUESTED** [review]
## FlagGems Automated Review

**Operator:** `_thnn_fused_lstm_cell`
**Verdict:** REQUEST_CHANGES

### Summary
- Errors: 1
- Warnings: 1
- Suggestions: 0

---
<sub>Automated review by FlagGems PR Review Skill</sub>
